features/prog.go

```diff
@@ -186,69 +170,57 @@ func HaveProgramHelper(pt ebpf.ProgramType, helper asm.BuiltinFunc) (err error)
 		err = wrapProbeErrors(err)
 	}()
 
-	if err := validateProgramType(pt); err != nil {
-		return err
-	}
-
-	if err := validateProgramHelper(helper); err != nil {
-		return err
-	}
-
-	return haveProgramHelper(pt, helper)
-}
-
-func validateProgramHelper(helper asm.BuiltinFunc) error {
 	if helper > helper.Max() {
 		return os.ErrInvalid
 	}
 
-	return nil
+	return haveProgramHelper(pt, helper)
 }
 
 func haveProgramHelper(pt ebpf.ProgramType, helper asm.BuiltinFunc) error {
-	pc.helperMu.Lock()
-	defer pc.helperMu.Unlock()
-	if err, ok := pc.helpers[pt][helper]; ok {
+	helperCache.Lock()
+	defer helperCache.Unlock()
+
+	key := helperKey{pt, helper}
+	if err, ok := helperCache.results[key]; ok {
 		return err
 	}
 
-	attr, err := createProgLoadAttr(pt, helper)
-	if err != nil {
-		return fmt.Errorf("couldn't create the program load attribute: %w", err)
+	if err := HaveProgramType(pt); err != nil {
```

Comment:
This is a change in behaviour: when probing for a helper we always probe the program type. Opinions?

Reply:
What is the benefit of doing it this way considering we call `probeProgram(spec)` both times?

Reply:
Good question, I can't remember myself xD As best as I can tell it has the following advantages:

- It returns a "program type not supported" instead of "helper for program type not supported" on old kernels. Tiny change, but maybe less confusing for users.
- It avoids having to handle E2BIG in this function, since HaveProgramType takes care of that.